[Enemy] Add ghost & demon implementations

tanwayne890 opened the pull request:
Closes #213 

This PR adds:
- add invulnerabilities and spawn ability to `EnemyInfo`
- add Ghost to invulnerable for 3 seconds (red sphere surrounds) and vulnerable for 5 seconds
- add Demon to spawn 3 ghost upon death with separate distance
- add GhostLoadout & GhostDemonScene in `dev` for play testing

video sent to tele (too large to send here):
- ghost immunes to damage and effects when red sphere appears
- demon spawns 3 ghost upon death, enemy number increases as well (see top right)

Should also test on every enemy prefab to see if working.

Note for art team @Eriksen2411: for every enemy prefab, the model name is specific: "Model". otherwise will not work properly
![image](https://user-images.githubusercontent.com/58804560/193798596-05158e0b-9ab8-4563-8ebb-f9bf8f5c583e.png)

tanwayne890 committed:
rename getStatus and status

tanwayne890 committed:
add invulnerabilities into EnemyInfo

tanwayne890 committed:
add ghost level

tanwayne890 committed:
stop apply effect when invulnerable

tanwayne890 committed:
remove slowspeed

tanwayne890 committed:
add ghost level to spawn ghost and demon

tanwayne890 committed:
add demon and mechanics

tanwayne890 commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -3,44 +3,71 @@
- */
-public enum EffectStatus {
+
+public enum Status {
+    /**
+     * Element status
+     */
     SCORCH, // fire
     CHILL, // ice
     DRENCH, // water
     SCALD, // fire + water
     FROZE, // ice + water
     WEAKEN, // fire + ice
+
+    /**
+     * Unique status
+     */
+    INVUlNERABLE, // immune to everything
     NONE // default
 }
```
previously was `EffectStatus`, but now is `Status` contains effect status and unique status (invulnerable)

tanwayne890 commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -154,6 +185,26 @@ public class Enemy : MonoBehaviour {
     }
 
     private void Die() {
+        if (isSpawnable && spawnCount > 0 && spawnPrefab != null) {
+            for (int i = 0; i < spawnCount; i++) {
+                GameObject spawnGO = (GameObject)Instantiate(spawnPrefab, transform.position, Quaternion.identity);
+
+                /*
+                 * set movement
+                 */
+                Enemy enemyScript = spawnGO.GetComponent<Enemy>();
+                if (enemyScript != null) {
+                    enemyScript.InitWaypoint(waypointIndex);
+                }
+
+                /*
+                 * wave enemy number
+                 */
+                WaveSpawner.numEnemiesLeftInWave++;
+                WaveSpawner.numEnemiesAlive++;
+            }
+        }
```
When the enemy with spawn ability dies, spawn the mobs. Here need to set the current `waypoint index` to the new mobs for movement

tanwayne890 commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -195,18 +250,34 @@ public class Enemy : MonoBehaviour {
             Die();
         }
 
+        if (isInvulnerable && GetStatus() == Status.INVUlNERABLE) {
+            if (duration <= 0f) {
+                DisableInvulnerability();
+            }
+            duration -= Time.deltaTime;
+        }
+
+        if (isInvulnerable && GetStatus() != Status.INVUlNERABLE) {
+            if (cooldown <= 0f) {
+                EnableInvulnerability();
+            }
+            cooldown -= Time.deltaTime;
+        }
```
invulnerable has a `duration`. after the duration, enemy becomes vulnerable for a period of time which is `cooldown`

tanwayne890 committed:
seperate mob spawn

tanwayne890 committed:
remove comments

tanwayne890 commented:
Added a fix to separate between the spawn. Instead of spawning them in fixed offset, use random values

tanwayne890 committed:
fix spawn in base

tanwayne890 commented:
ready for review!

trxe commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -3,44 +3,70 @@
+    /**
+     * Unique status
+     */
+    INVUlNERABLE, // immune to everything
```
```suggestion
    INVULNERABLE, // immune to everything

```

tanwayne890 replied:
https://github.com/CS3247-Game-Development-Team-6/Doodles/pull/217/commits/5f4df28c433932577272dc39212a71b7c907e112

trxe commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -281,4 +379,21 @@ public class Enemy : MonoBehaviour {
+        cooldown = enemyInfo.cooldown;
+
+        // only for invulnerable enemy
+        transform.GetChild(4).gameObject.SetActive(true);
```
two issues:

1. visible through fog:

![image](https://user-images.githubusercontent.com/23572486/192130233-f33ddf98-3529-4504-bcd0-6e68d7099cc4.png)

2. hardcoding:

for now, instead of hardcoding the child index, search for the gameobject by name or keep a private reference to it please

tanwayne890 replied:
both fixed. no.1 in https://github.com/CS3247-Game-Development-Team-6/Doodles/pull/217/commits/5b4038e7cff42846f7befc73c4c7027e1d6d054c
no.2 in https://github.com/CS3247-Game-Development-Team-6/Doodles/pull/217/commits/4ae8743b648a100088947f78cccf201a5fc62eea

trxe commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -281,4 +379,21 @@ public class Enemy : MonoBehaviour {
+        duration = enemyInfo.duration;
+
+        // only for invulnerable enemy
+        transform.GetChild(4).gameObject.SetActive(false);
```
same as above

trxe commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -153,7 +183,47 @@ public class Enemy : MonoBehaviour {
+    /**
+     * offset to separate between mobs
+     */
+    private void SpawnWhenDeath(GameObject _prefab, Vector3 _spawnPosition, Vector3 minRange, Vector3 maxRange, int _spawnCount) {
```
If we have time, we should discuss how to make Enemy a parent class and all implementations of Enemy such as this one inherit from Enemy instead. If not the enemy class will be cluttered with the methods for all the types of Enemies we make

tanwayne890 replied:
Idea similar to tower? like EnemyInfo <- (implements) Enemy <- (extends) CustomEnemy

After I developed laser tower using our tower framework, probably have better ideas to implement enemy

trxe reviewed (changes requested):
Please fix the suggested points, and after tonight's call we can discuss a bit on abstracting out the Enemy functions if possible.

trxe commented on Assets/Scripts/Enemy/EnemyInfo.cs:
```diff
@@ -26,5 +26,20 @@ public class EnemyInfo : ScriptableObject {
+     * Spawnable after death
+     */
+    [Header("Spawnable")]
+    public bool isSpawnable;
```
```suggestion
    public bool spawnsOnDeath;
```

tanwayne890 replied:
https://github.com/CS3247-Game-Development-Team-6/Doodles/pull/217/commits/31a35cbca8e02543cbdc66d6c734fb2d5b2b3607

tanwayne890 committed:
Merge branch 'main-v2' into ghost-demon

# Conflicts:
#	Assets/Scripts/ElementEffectSystem/EffectManager.cs
#	Assets/Scripts/Enemy/Enemy.cs
#	Assets/Scripts/GameManager/WaveSpawner.cs
#	Assets/Scripts/Map/Waypoints.cs

tanwayne890 committed:
Add ghostloadout, ghostdemonscene and ghostchunk

tanwayne890 committed:
cleanup

tanwayne890 committed:
typo invulnerable

tanwayne890 committed:
isSpawnable -> spawnOnDeath

tanwayne890 committed:
getChild -> FInd(stringName)

tanwayne890 committed:
require "Model" name

tanwayne890 committed:
remove unused

trxe reviewed (approved):
Is working, but may need to revamp the Enemy implementation in the future as it is getting cluttered.

tanwayne890 committed:
add demon death and spawn effects

tanwayne890 commented:
Waiting for ghost and demon model

Eriksen2411 committed:
Add model and animation for lvl 2 enemies

- Add models and animations for normal ghost, invulnerable ghost and demon.
- Add Normal ghost implementation.
- Currently playtest in GhostDemonScene scene.
